### ISSUE TITLE
Kernel 5.17.4 on Thinkpad W520 - no audio (5.16.18-2 works)

### ISSUE THREAD

**Issue by brendanhoar** [QubesOS/qubes-issues#7465]
### Qubes OS release
R4.0 kernel-latest current-testing

### Brief summary
Thinkpad W520
Audio works with kernel 5.16.18-2
Audio does not work with kernel 5.17.4

### Steps to reproduce
Update W520 running R4.0 to kernel-latest in current-testing repo.

Notably, I see the following repeated in dmesg under 5.17.4, but I do not see this in dmesg in 5.16.18-2:

```
...snd_hda_intel swiotlb buffer is full ...
```

### Expected behavior
Audio

### Actual behavior
No audio
Reverting to 5.16.18-2 during boot temporarily fixes the issue.

**Comment by brendanhoar** [comment]
Newly released 5.17.5 seems to have resolved the issue on my W520.

Tested two different reboots so far:

1. Audio works.
2. No more swiotlb errors related to the sound hardware found in dom0's dmesg output.

Unclear from kernel change-set what might have fixed the issues (if anything).

B

**Comment by Minimalist73** [comment]
Kernel 5.17.5 didn't fix the issue on my side, the sound just stopped after a few hours of use. I'm not getting that issue with kernel 5.16.18.

**Comment by brendanhoar** [comment]
Hmm ok I'll keep testing.

**Comment by DemiMarie** [comment]
@Minimalist73 @brendanhoar would you mind reporting the bugs to regressions@lists.linux.dev?

**Comment by brendanhoar** [comment]
> Kernel 5.17.5 didn't fix the issue on my side, the sound just stopped after a few hours of use. I'm not getting that issue with kernel 5.16.18. Nothing useful in logs too or related to audio in dmesg.

Confirmed here as well.

With 5.17.5-1 in dom0 (R4.0) on Thinkpad W520:

aplay -q ./sinewave.wav --duration=1 

works immediately after reboot but later fails with:

ALSA lib pcm_dmix.c:1029:(snd_pcm_dmix_open) unable to open slave
aplay: main:786: audio open error: Device or resource busy

B

**Comment by marmarek** [comment]
I can confirm the issue as well. After some hours of uptime audio stops working. I don't see `snd_hda_intel swiotlb buffer is full` message in my case, though.

> aplay -q ./sinewave.wav --duration=1

`aplay` is expected to fail, device is grabbed by pulseaudio. But you can try `paplay`.

**Comment by DemiMarie** [comment]
> After some hours of uptime audio stops working.

Ouch.  That is going to make bisecting hard.

**Comment by brendanhoar** [comment]
> aplay is expected to fail, device is grabbed by pulseaudio. But you can try paplay

Thanks. 

aplay works after boot *until* device is grabbed by pulseaudio.

So if I use paplay in dom0 once after reboot, aplay fails for a bit but starts working again a few moments later (I presume PA relinquished the device). 

aplay is useful for initial testing dom0 via script because it errors out while paplay does not.

I'm going to try some diagnosis starting with no VMs running to see how long it takes before sound fails/aplay throws error/swiotlb errors appear (assuming it fails).

If I can't trigger it that way, then I'll try running some non-pci-attached VMs and repeat dom0 tests, though I may need to work something out with paplay at that point.

Similarly if that doesn't produce errors, I'll add some PCI-attached VMs to the mix and retest.

B

PS - earlier I found that when the sound was broken, sometimes starting up other VMs triggered producing a looping semi random noise (well, likely non-sound data being sampled/looped by the hardware). Unclear what triggered that. Scared my older kid, sounded awful. :)

**Comment by brendanhoar** [comment]
Switched to paplay once VMs are running.

No problems reported while running a vault VM for a couple hours.

Started sys-net and dom0 sound died.

Shut down sys-net and dom0 sound started working again.

Started sys-net and dom0 sound...kept working.

Starting experimenting with others.

Hard to say but starting and shutting down VMs seems like it can trigger temporary (30s) or semi-permanent (until. m is shut down) sound failures in dom0. I always see the swiotlb errors in dom0 when this happens.

Failure/errors *seem* potentially correlated with initial qmemman rebalancing, still testing.

B

**Comment by marmarek** [comment]
One message I noticed:
```
[ 2142.484553] snd_hda_intel 0000:00:1f.3: Unstable LPIB (18144 >= 6396); disabling LPIB delay counting
```
I'm not sure about the timing, but when I noticed audio stopped working, this message was logged some time before it.

**Comment by brendanhoar** [comment]
So far today I couldn't get audio to stop working after setting the /etc/default/grub min and max dom0 memory to 2048M running sudo grub2-mkconfig -o /boot/grub2/grub.cfg and then rebooting.

Experimented with running enough memory balancing VMs on a 32GB W520 such that I could no longer start any new VMs, shut them all down, tried again, etc. 

I did see one swiotlb error early on but nothing past that and audio continued to work the whole time.


B

**Comment by marmarek** [comment]
Another information: unloading all `snd*` modules and loading them again does _not_ fix audio.

**Comment by brendanhoar** [comment]
Well my testing has proven unproductive. It appears that playing wav files repeatedly in dom0 via paplay seems to itself significantly reduce the chances of triggering the symptom. If I don't do that then the symptom returns much more often even with dom0 set to a static memory amount.

Another direction to go might be attempting to increase log levels somewhere? Pulseaudio? ALSA? Driver?

B

**Comment by DemiMarie** [comment]
Can anyone do a bisection?  This is obviously a regression, but without a bisection it likely won’t be fixed.

**Comment by marmarek** [comment]
> Can anyone do a bisection?

I don't think that's realistic, given observations by @brendanhoar. I have similar observation - it stopped working for me yesterday only after over a day uptime. We don't know how long it needs to work to consider the commit not broken.

**Comment by DemiMarie** [comment]
> > Can anyone do a bisection?
> 
> I don't think that's realistic, given observations by @brendanhoar. I have similar observation - it stopped working for me yesterday only after over a day uptime. We don't know how long it needs to work to consider the commit not broken.

Can it be parallelised in any way?

**Comment by marmarek** [comment]
> Can it be parallelised in any way?

git bisect? no, because the next step depends on previous one; and also, we (collectively) don't have enough systems identified to be affected - it seems right now we have two: one @brendanhoar and one mine

**Comment by brendanhoar** [comment]
Since:

a) we seem to have an external standard Linux (non-qubes) case in that forum thread.

b) the fault often seems to occur right at the moment that qubes sends the notification that a VM has started and 

c) increasing pulseaudio verbosity shows some (non-error) messages around that notification  about sample rate changes and or idle transitions in the target sink

...I suspect we might be able to create some test scripts in dom0 to attempt to trigger the issues.

Just a feeling.

B

**Comment by brendanhoar** [comment]
I definitely can trigger the issues just with dom0 playing  several wav files of 8k, 44.1k and 48k sampling rates using paplay. However it is still unclear the exact steps needed to trigger and clear the issue.

Running the below:

sudo watch -n 0.1 -d 'cat /proc/asound/card0/pcm0*/sub0/hw_params'

I can watch out for sample rate changes, either via dom0 wav files or startup of VMs and or audio in VMs.

Notable sometimes the hw_params returns "closed" after some short timeout with no audio, other times it says stuck open (not sure why).

I've wanted to do some traces on the snd_hda_intel driver but it appears the ability to enable it isn't available, as the following path is missing:

/sys/kernel/debug/tracing/events/hda/

So I can't echo 1>enabled in that dir.

Anyone with experience: is the reason the hda folder is missing from debugfs because the Qubes kernel does not have CONFIG_SND_DEBUG=y...or is there another reason?

Thanks,
Brendan

**Comment by pgerber** [comment]
I'm also experiencing audio issue on a ThinkPad running Qubes 4.1 ever since updating to 5.17. Pitches are off in combination with short cut outs but sometimes audio also doesn't work at all. So far the issue seems rather reproducible and I don't have to wait for the issue appear, it's just there.

I do see some warning in the kernel logs which may help here:

```
May 13 17:16:25 dom0 kernel: ------------[ cut here ]------------
May 13 17:16:25 dom0 kernel: i915 0000:00:02.0: drm_WARN_ON(dig_port->tc_mode != TC_PORT_DISCONNECTED)
May 13 17:16:25 dom0 kernel: WARNING: CPU: 5 PID: 74 at drivers/gpu/drm/i915/display/intel_tc.c:705 intel_tc_port_sanitize+0x1b9/0x380 [i915]
May 13 17:16:25 dom0 kernel: Modules linked in: loop vfat fat snd_ctl_led snd_soc_skl_hda_dsp snd_soc_intel_hda_dsp_common snd_soc_hdac_hdmi snd_hda_codec_hdmi snd_hda_codec_realtek snd_hda_codec_generic snd_soc_dmic snd_sof_pci_intel_tgl snd_sof_intel_>
May 13 17:16:25 dom0 kernel:  processor_thermal_rfim processor_thermal_mbox processor_thermal_rapl intel_rapl_common intel_soc_dts_iosf ucsi_acpi typec_ucsi typec wmi thinkpad_acpi platform_profile ledtrig_audio rfkill snd soundcore int3403_thermal int3>
May 13 17:16:25 dom0 kernel: CPU: 5 PID: 74 Comm: kworker/u16:1 Tainted: G        W         5.17.7-1.fc32.qubes.x86_64 #1
May 13 17:16:25 dom0 kernel: Hardware name: LENOVO 20Y3CTO1WW/20Y3CTO1WW, BIOS N40ET31W (1.13 ) 11/23/2021
May 13 17:16:25 dom0 kernel: Workqueue: events_unbound async_run_entry_fn
May 13 17:16:25 dom0 kernel: RIP: e030:intel_tc_port_sanitize+0x1b9/0x380 [i915]
May 13 17:16:25 dom0 kernel: Code: 4c 8b 6f 50 4d 85 ed 75 03 4c 8b 2f e8 30 1b 5a c1 48 c7 c1 c0 b2 47 c0 4c 89 ea 48 c7 c7 50 a8 49 c0 48 89 c6 e8 1f b6 99 c1 <0f> 0b 41 8b 8f a0 0f 00 00 85 c9 0f 84 b4 fe ff ff 48 8b 7b 08 4c
May 13 17:16:25 dom0 kernel: RSP: e02b:ffffc90040343c58 EFLAGS: 00010282
May 13 17:16:25 dom0 kernel: RAX: 0000000000000000 RBX: ffff88810b900000 RCX: 0000000000000000
May 13 17:16:25 dom0 kernel: RDX: 0000000000000001 RSI: ffffffff826835d3 RDI: 00000000ffffffff
May 13 17:16:25 dom0 kernel: RBP: 0000000000000000 R08: 0000000000000000 R09: 00000000834f49aa
May 13 17:16:25 dom0 kernel: R10: ffffffffffffffff R11: 0000000000000005 R12: ffff88810b14cf80
May 13 17:16:25 dom0 kernel: R13: ffff8881011cea80 R14: ffff88810b9002a8 R15: ffff88810b14c000
May 13 17:16:25 dom0 kernel: FS:  0000000000000000(0000) GS:ffff888171b40000(0000) knlGS:0000000000000000
May 13 17:16:25 dom0 kernel: CS:  e030 DS: 0000 ES: 0000 CR0: 0000000080050033
May 13 17:16:25 dom0 kernel: CR2: 00005bfdd3c3ce86 CR3: 0000000002810000 CR4: 0000000000050660
May 13 17:16:25 dom0 kernel: Call Trace:
May 13 17:16:25 dom0 kernel:  <TASK>
May 13 17:16:25 dom0 kernel:  ? ktime_get_mono_fast_ns+0x46/0x90
May 13 17:16:25 dom0 kernel:  intel_ddi_sync_state+0x40/0x90 [i915]
May 13 17:16:25 dom0 kernel:  intel_modeset_readout_hw_state+0x193/0x770 [i915]
May 13 17:16:25 dom0 kernel:  intel_modeset_setup_hw_state+0x4a/0x3c0 [i915]
May 13 17:16:25 dom0 kernel:  ? drm_modeset_lock_all_ctx+0x226/0x350
May 13 17:16:25 dom0 kernel:  intel_display_resume+0x96/0x140 [i915]
May 13 17:16:25 dom0 kernel:  ? drm_dp_mst_topology_mgr_set_mst+0x254/0x3c0
May 13 17:16:25 dom0 kernel:  i915_drm_resume+0xbf/0x140 [i915]
May 13 17:16:25 dom0 kernel:  ? pci_pm_poweroff_noirq+0x110/0x110
May 13 17:16:25 dom0 kernel:  dpm_run_callback+0x49/0x130
May 13 17:16:25 dom0 kernel:  device_resume+0xa7/0x200
May 13 17:16:25 dom0 kernel:  async_resume+0x19/0x30
May 13 17:16:25 dom0 kernel:  async_run_entry_fn+0x2d/0x130
May 13 17:16:25 dom0 kernel:  process_one_work+0x1e2/0x3b0
May 13 17:16:25 dom0 kernel:  worker_thread+0x49/0x2e0
May 13 17:16:25 dom0 kernel:  ? rescuer_thread+0x3a0/0x3a0
May 13 17:16:25 dom0 kernel:  kthread+0xe5/0x110
May 13 17:16:25 dom0 kernel:  ? kthread_complete_and_exit+0x20/0x20
May 13 17:16:25 dom0 kernel:  ret_from_fork+0x1f/0x30
May 13 17:16:25 dom0 kernel:  </TASK>
May 13 17:16:25 dom0 kernel: ---[ end trace 0000000000000000 ]---
May 13 17:16:25 dom0 kernel: ------------[ cut here ]------------
```

[logs.txt](https://github.com/QubesOS/qubes-issues/files/8691551/logs.txt)

**Comment by marmarek** [comment]
Some more confusing observations:
1. When I detach pulseaudio from the device, I can use `aplay something.wav` to play audio (same file and format each time, I believe the exact format doesn't really matter) - it will _sometimes_ work, it looks like waiting a little between trials makes it more likely it will work. But trying one immediately after another often works too. 
2. When connecting pulseaudio back to the device, most of the time it does not work. But if I wait for pulseaudio to suspend the output (watch `pactl list sinks`) which I think means it closes the device, then after waiting a bit it starts working again. Nothing can play nor monitor the output ("sink") for pulseaudio to suspend it, which includes all the VMs but also you need to close pavucontrol or similar apps.

No kernel module reloading is involved here.

---
If you have pipewire in dom0 running, kill it, or even uninstall it, it will only get in the way.
Convincing pulseaudio to detach from device: `pactl unload-module module-alsa-card`
Connecting pulseaudio to the device again: `pactl unload-module module-udev-detect; pactl load-module module-udev-detect`

**Comment by brendanhoar** [comment]
It's rather maddening. 

I haven't had a block of free time to continue diagnosis since my last post.

B

**Comment by DemiMarie** [comment]
@brendanhoar @marmarek looks like it will take a kernel bisection.

**Comment by marmarek** [comment]
> @brendanhoar @marmarek looks like it will take a kernel bisection.

https://github.com/QubesOS/qubes-issues/issues/7465#issuecomment-1120904886

**Comment by Minimalist73** [comment]
To add more to this issue, I still get audio problems with last kernel (5.17.7) but I have found another issue as well. From time to time, when I attach my mic to a qube (jack headset with splitter), voice don't work at all in qube but I can see the bar move when I open pavucontrol in dom0, it's like something between is broken. I'll try to reproduce this issue when I have time and I'll check logs for anything useful.

**Comment by DemiMarie** [comment]
@pgerber: Can you do a kernel bisection?  Your problem appears to reproduce immediately after boot.

**Comment by pgerber** [comment]
> @pgerber: Can you do a kernel bisection? Your problem appears to reproduce immediately after boot.

@DemiMarie, I built the kernel using the config from dom0 but ended up with a kernel that was unbootable. That is, I'd just see a black screen for a few seconds before the machine reset. Is there a set of instructions for building a dom0 kernel somewhere I can follow?

**Comment by pgerber** [comment]
Yet another piece of information, turns out when I'm running 5.17, opening the Audio mixer (audio tray → Audio mixer…), sound works just fine. When I close the mixer, sound degrades again within ~10 seconds and when I open it returns back to normal almost immediately.

**Comment by DemiMarie** [comment]
> > @pgerber: Can you do a kernel bisection? Your problem appears to reproduce immediately after boot.
> 
> @DemiMarie, I built the kernel using the config from dom0 but ended up with a kernel that was unbootable. That is, I'd just see a black screen for a few seconds before the machine reset. Is there a set of instructions for building a dom0 kernel somewhere I can follow?

Not that I am aware of, sadly.  The best option I know of is to change the builder to fetch and verify a different kernel version, then use that.

**Comment by marmarek** [comment]
> @DemiMarie, I built the kernel using the config from dom0 but ended up with a kernel that was unbootable. That is, I'd just see a black screen for a few seconds before the machine reset. Is there a set of instructions for building a dom0 kernel somewhere I can follow?

Do you have initramfs for this kernel? If not, try to create it with `dracut --kver (kernel-version)` (get kernel version with `make kernelrelease` in the linux sources). Then, update grub as usual.

**Comment by Minimalist73** [comment]
Little update, I lost sound again with the latest kernel available (5.18.3-1), after trying a few things and finding no error at all related to sound in dmesg, I fully restarted pulseaudio with the following command:

```
systemctl restart --user pulseaudio*
```
After execution of it, sound came back fully but input sound is still broken (bar don't even move in pulseaudio while talking).

**Comment by pgerber** [comment]
> Do you have initramfs for this kernel? If not, try to create it with dracut --kver (kernel-version) (get kernel version with make kernelrelease in the linux sources). Then, update grub as usual.

I built and RPM package which handles initramfs creation on install.

> Not that I am aware of, sadly. The best option I know of is to change the builder to fetch and verify a different kernel version, then use that.

I've tried using Qubes Builder as described in https://www.qubes-os.org/doc/qubes-builder/ and I tried it to just build Qubes unmodified first but I ended up with this error:

```
--> Downloading additional sources for vmm-xen-stubdom-legacy...
make[1]: Entering directory '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy'
git submodule update --init --recursive
Submodule 'core-vchan-xen' (git://github.com/QubesOS/qubes-core-vchan-xen) registered for path 'core-vchan-xen'
Submodule 'gui-agent-xen-hvm-stubdom' (git://github.com/QubesOS/qubes-gui-agent-xen-hvm-stubdom) registered for path 'gui-agent-xen-hvm-stubdom'
Submodule 'gui-common' (git://github.com/QubesOS/qubes-gui-common) registered for path 'gui-common'
Cloning into '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/core-vchan-xen'...
fatal: unable to connect to github.com:
github.com[0: 140.82.121.4]: errno=Connection timed out

fatal: clone of 'git://github.com/QubesOS/qubes-core-vchan-xen' into submodule path '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/core-vchan-xen' failed
Failed to clone 'core-vchan-xen'. Retry scheduled
Cloning into '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/gui-agent-xen-hvm-stubdom'...
fatal: unable to connect to github.com:
github.com[0: 140.82.121.4]: errno=Connection timed out

fatal: clone of 'git://github.com/QubesOS/qubes-gui-agent-xen-hvm-stubdom' into submodule path '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/gui-agent-xen-hvm-stubdom' failed
Failed to clone 'gui-agent-xen-hvm-stubdom'. Retry scheduled
Cloning into '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/gui-common'...
fatal: unable to connect to github.com:
github.com[0: 140.82.121.4]: errno=Connection timed out

fatal: clone of 'git://github.com/QubesOS/qubes-gui-common' into submodule path '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/gui-common' failed
Failed to clone 'gui-common'. Retry scheduled
Cloning into '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/core-vchan-xen'...
fatal: unable to connect to github.com:
github.com[0: 140.82.121.3]: errno=Connection timed out

fatal: clone of 'git://github.com/QubesOS/qubes-core-vchan-xen' into submodule path '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy/core-vchan-xen' failed
Failed to clone 'core-vchan-xen' a second time, aborting
make[1]: *** [Makefile:56: get-sources] Error 1
make[1]: Leaving directory '/home/user/src/qubes-build/qubes-builder/qubes-src/vmm-xen-stubdom-legacy'
make: *** [Makefile:226: vmm-xen-stubdom-legacy.get-sources-extra] Error 2
```

Looks like the builder tries to fetch some resources via git:// scheme which I don't think is even supported by GitHub.

**Comment by marmarek** [comment]
Looks to be an older commit, should be better now. Anyway, to build just the kernel, you can modify `builder.conf` to include only `linux-kernel` in COMPONENTS, and enable https://www.qubes-os.org/doc/qubes-builder/#use-pre-built-qubes-packages for dependencies. It will be much quicker than downloading everything.

**Comment by marmarek** [comment]
`git bisect` (that took almost 2 months...) said:
```
2c95b92ecd92e784785b1db8cccc4f0f2bfa850c is the first bad commit
commit 2c95b92ecd92e784785b1db8cccc4f0f2bfa850c
Author: Takashi Iwai <tiwai@suse.de>
Date:   Tue Nov 16 08:33:58 2021 +0100

    ALSA: memalloc: Unify x86 SG-buffer handling (take#3)
    
    This is a second attempt to unify the x86-specific SG-buffer handling
    code with the new standard non-contiguous page handler.
    
    The first try (in commit 2d9ea39917a4) failed due to the wrong page
    and address calculations, hence reverted.  (And the second try failed
    due to a copy&paste error.)  Now it's corrected with the previous fix
    for noncontig pages, and the proper sg page iteration by this patch.
    
    After the migration, SNDRV_DMA_TYPE_DMA_SG becomes identical with
    SNDRV_DMA_TYPE_NONCONTIG on x86, while others still fall back to
    SNDRV_DMA_TYPE_DEV.
    
    Tested-by: Alex Xu (Hello71) <alex_y_xu@yahoo.ca>
    Tested-by: Harald Arnesen <harald@skogtun.org>
    Link: https://lore.kernel.org/r/20211017074859.24112-4-tiwai@suse.de
    Link: https://lore.kernel.org/r/20211109062235.22310-1-tiwai@suse.de
    Link: https://lore.kernel.org/r/20211116073358.19741-1-tiwai@suse.de
    Signed-off-by: Takashi Iwai <tiwai@suse.de>

 include/sound/memalloc.h |  14 ++--
 sound/core/Makefile      |   1 -
 sound/core/memalloc.c    |  53 ++++++++++++-
 sound/core/sgbuf.c       | 201 -----------------------------------------------
 4 files changed, 56 insertions(+), 213 deletions(-)
 delete mode 100644 sound/core/sgbuf.c
```

The fact that it's Nth attempt already doesn't build confidence, and reviewing git log I see few later fixups for this already. The most recent that mentions the above commit explicitly is [925ca893b4a](https://github.com/torvalds/linux/commit/925ca893b4a), but it's included in 5.18 already, and there were reports of this issue happening with 5.18 too. But there are also few more kind of related commits in 5.19, so my next step is to try 5.19.12 for some time.

**Comment by Minimalist73** [comment]
I'm currently using kernel 6.0.2 and I lost sound after 9h of use today. I fixed it by changing profile in pavucontrol from `Analog Stereo Duplex` to `Analog Surround 7.1 Output` (only this one works after the issue). 1h after the issue, I tried to come back to `Analog Stereo Duplex` and it worked when it wasn't before.

**Comment by marmarek** [comment]
Upstream report: https://lore.kernel.org/alsa-devel/Y6E9VixS7sV0kpWQ@mail-itl/T/
